src/protocols/secure_channel/CASESession.cpp

```diff
@@ -144,28 +144,74 @@ void CASESession::Clear()
     mState = State::kInitialized;
     Crypto::ClearSecretData(mIPK);
 
+    // TODO we need to use the fabric table provided after Tennessee's PR.
+    //mFabricTable->RemoveFabricDelegate(&mFabricDelegate);
+
     mLocalNodeId = kUndefinedNodeId;
     mPeerNodeId  = kUndefinedNodeId;
     mFabricInfo  = nullptr;
 }
 
+void CASESession::InvalidateIfPendingEstablishment()
+{
+    // TODO reconsider if we need to include all states. It is very unlikely that any
+    // new states will be added so it could be say to early exit on states that don't
+    // need any cleanup.
+    switch (mState)
+    {
+    case State::kSentSigma1:
+        FALLTHROUGH;
+    case State::kSentSigma2:
+        FALLTHROUGH;
+    case State::kSentSigma3:
+        FALLTHROUGH;
+    case State::kSentSigma1Resume:
+        FALLTHROUGH;
+    case State::kSentSigma2Resume:
+        // TODO #1 Is there anything else we need to do here. This delegate technically
+        // calls Clear() for use, but I don't think that is guaranteed, just how it
+        // is implemented today.
+        // TODO #2 double check if there is maybe a more suitable CHIP_ERROR
+        mDelegate->OnSessionEstablishmentError(CHIP_ERROR_INCORRECT_STATE);
+        break;
+    case State::kInitialized:
+        // TODO double check that for kInitialized that we don't want to also tear things
+        // down.
+        FALLTHROUGH;
+    case State::kFinished:
+        FALLTHROUGH;
+    case State::kFinishedViaResume:
+        break;
+    }
+}
+
 CHIP_ERROR CASESession::Init(SessionManager & sessionManager, SessionEstablishmentDelegate * delegate)
 {
+    CHIP_ERROR err = CHIP_NO_ERROR;
     VerifyOrReturnError(delegate != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
     VerifyOrReturnError(mGroupDataProvider != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
 
     Clear();
 
+    // No need to run SuccessOrExit as 
     ReturnErrorOnFailure(mCommissioningHash.Begin());
 
+    // TODO we need to use the fabric table provided after Tennessee's PR.
+    // SuccessOrExit(err = mFabricTable->AddFabricDelegate(&mFabricDelegate));
```

Comment:
If we don't want to `AddFabricDelegate` and `RemoveFabricDelegate` called for every CASESession establishment (which is what I am starting to think is where I will take this PR), we could use SessionManager. We can piggy back on the `FabricTable::Delegate`already implemented in `Server.h` that calls into the singleton SessionManager. We would just register CASESession that are in the process of being established.